### PR TITLE
dex.offers 🦖

### PR DESCRIPTION
**Pull request by denverbaumgartner**
Brief comments on the purpose of your changes: 

short purpose: 
we currently have `dex.trades` but no `dex.offers`, this changes that! the initial motivation was to enable easy cross chain data / aggregation for Rubicon, a fully on-chain order book (and currently the only DEX in `dex.offers`). 

longer purpose: 
this spell sets a foundation for other order book DEXs (as they emerge) to contribute to. as well, i believe this could be a way to track historical execution prices across all DEXs. the approach would not be as simple when it comes to an AMM (the goal here being "what rate could i have swapped USDC -> ETH on xyz AMM at block 1,000,000), but i think its definitely feasible given some constraints (most likely setting some type of bin size to model an order book off of an AMM's constant function (ex: solve for the scenario previously mentioned (USDC -> ETH at block 1,000,000) by calculating the cost of purchasing ETH in .1 increments (0.1, 0.2, ..., 5.1, etc.)). i hope to take a stab at this sometime in the near future, but for now im off to fight other 🔥s 

some general thoughts: 

- i wanted to be as chain agnostic as possible, so transaction costs are just represented as USD values (transaction costs being the primary limiter of on chain order books (s/o L2 scaling solutions 🙏 ))
- going through the dex tests, it looks like they mainly depend on the lower level tables being well tested, rather than having extensive testing at the aggregate level. so, `rubicon_optimism_offers` has full test coverage, and ideally that sound state trickles up to `dex.offers` 💯  *any feedback here is appreciated 🤹 
- within the `dbt_project.yml` file i added in `dex.offers` in a similar fashion to the other aggregate tables, but it looks like i was more verbose with types defined for every column, wasn't sure why others hadn't done this, but expect i will find out as a result of this pr ;)

would love any thoughts / feedback on how to improve the work done here so far. i await everyone's feedback and am excited to do my small part to support the ecosystem 🖖 

**For Dune Engine V2**

I've checked that:

### General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [x] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [x] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

File: models/rubicon/optimism/rubicon_optimism_offers.sql
Comment (Hosuke) [inline models/rubicon/optimism/rubicon_optimism_offers.sql:203]: Hi @denverbaumgartner, we may need to fix the cast to varchar here.
```suggestion
    AND cast(eth.contract_address AS varchar(100)) = '0xdeaddeaddeaddeaddeaddeaddeaddeaddead0000' -- this is for optimism specifically
```

Comment (denverbaumgartner) [reply models/rubicon/optimism/rubicon_optimism_offers.sql]: hey @Hosuke, just fixed this, thanks for the suggestion 💯

File: tests/rubicon/optimism/rubicon_optimism_offers_test.sql
Comment (Hosuke) [inline tests/rubicon/optimism/rubicon_optimism_offers_test.sql:103]: There may be type parsing error in comparison when you compare `*_amount_raw`:
https://dune.com/queries/2267897?d=1

Comment (denverbaumgartner) [reply tests/rubicon/optimism/rubicon_optimism_offers_test.sql]: 👍

**Comment by denverbaumgartner** [comment]
> 

@Hosuke thanks for the feedback, i will split this into two PRs: 1) rubicon_offers update 2) dex.offers update. hopefully this will help!

File: models/dex/dex_offers.sql
Comment (Hosuke) [inline models/dex/dex_offers.sql:3]: For now we only keep this for optimism chain.
```suggestion
        post_hook='{{ expose_spells(\'["optimism"]\',
```

File: models/rubicon/optimism/rubicon_optimism_schema.yml
Comment (Hosuke) [inline models/rubicon/optimism/rubicon_optimism_schema.yml:108]: We can also apply a `check_seed` test here:
```suggestion
    tests:
      - check_seed:
          seed_file: ref('rubicon_optimism_offers_test_data')
          match_columns:
            - block_number
            - tx_hash
            - evt_index
            - tx_index
          check_columns:
            - l1_gas_used
            ...
      - dbt_utils.unique_combination_of_columns:
          combination_of_columns:
```

**Review by Hosuke — COMMENTED** [review]
Since I cannot directly commit to this PR, I will try to comment the suggestions here.

File: models/rubicon/optimism/rubicon_optimism_schema.yml
Comment (denverbaumgartner) [reply models/rubicon/optimism/rubicon_optimism_schema.yml]: implemented [here](https://github.com/duneanalytics/spellbook/pull/3042/commits/a364fd2746803871fde9794ca71fed037476393a), ty @Hosuke

**Review by Hosuke — APPROVED** [review]
LGTM.✅ for code logics.

Personally I think `dex.offers` is a great idea. Since this sector is still in early stage, we can still design the schema and implement this together.

You can allow the direct commits to this PR(and for future PRs) from maintainers of the spellbook, so we can commit to this PR easily.

We still need to pass all the CI tests before merging stage.

[Sample dex.offers](https://dune.com/queries/2296688)

**Comment by denverbaumgartner** [comment]
> LGTM.✅ for code logics.
> 
> Personally I think `dex.offers` is a great idea. Since this sector is still in early stage, we can still design the schema and implement this together.
> 
> You can allow the direct commits to this PR(and for future PRs) from maintainers of the spellbook, so we can commit to this PR easily.
> 
> We still need to pass all the CI tests before merging stage.
> 
> [Sample dex.offers](https://dune.com/queries/2296688)

@Hosuke it would be an honor to work on `dex.offers` with you and i am glad you like the idea 🙏 i am open to any and all suggestions you have and am excited to learn from your experience 

one the note of direct commits, it looks like this is actually an issue with GH preventing this feature for cross-organization PRs, see more [here](https://github.com/orgs/community/discussions/5634). in light of this, i will re-submit this PR from my personal account to allow for direct commits. 

thanks for sharing the sample `dex.offers` table, this actually highlights an error for me to correct! currently, the cost of the transactions is inflated as it is multiplying the cost of the transaction by the number of offers in the transaction (Rubicon allows you to "batch" offers, so multiple offers can be contained within the same transaction). i will correct for this in my updated PR 🖖

**Pull request closed without merging**
